### DISCUSSION TITLE
Event sampling from .xml

### DISCUSSION THREAD

**Discussion by astroSyvertHE** [gammapy/gammapy#4355] [Help]
How can I do event sampling from .xml? (gammapy1.0)

**Comment by bkhelifi** [comment]
Hello Sofia,
This is not supported, only the YAML format. Converters exit from XML to YAML.

**Comment by adonath** [comment]
Thanks @astroSyvertHE! Do you refer to the XML definition for the observations or the models? In both cases the answer is there is no direct support, but there are feasible workarounds. I'll ping @QRemy, as he has some experience with converting both.

**Reply by astroSyvertHE** [reply]
Yes, I have xml with models of sources to simulate. There are tutorials to simulate just one source compling the model parameters directly in notebook, but if I want to simulate many sources from xml? First I have to convert xml in yml, and then? thanks

**Reply by registerrier** [reply]
Once your models are defined in yaml file you can simply read it with `Models.read("model_file.yaml")`
See this section of the Model API tutorial from the documentation:
https://docs.gammapy.org/1.0/tutorials/api/models.html#model-lists-and-serialisation

**Reply by astroSyvertHE** [reply]
I'm trying to sample events from .yaml models with several sources. There's this overlap error. Could you help me? Thanks

---------------------------------------------------------------------------
NoOverlapError                            Traceback (most recent call last)
Input In [29], in <cell line: 2>()
      1 sampler = MapDatasetEventSampler(random_state=0)
----> 2 events = sampler.run(dataset, observation)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/datasets/simulate.py:358, in MapDatasetEventSampler.run(self, dataset, observation)
    341 """Run the event sampler, applying IRF corrections.
    342 
    343 Parameters
   (...)
    355     Event list.
    356 """
    357 if len(dataset.models) > 1:
--> 358     events_src = self.sample_sources(dataset)
    360     if len(events_src.table) > 0:
    361         if dataset.psf:

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/datasets/simulate.py:80, in MapDatasetEventSampler.sample_sources(self, dataset)
     71 if evaluator.needs_update:
     72     evaluator.update(
     73         dataset.exposure,
     74         dataset.psf,
   (...)
     77         dataset.mask,
     78     )
---> 80 flux = evaluator.compute_flux()
     81 npred = evaluator.apply_exposure(flux)
     83 if evaluator.model.temporal_model is None:

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/datasets/evaluator.py:240, in MapEvaluator.compute_flux(self, *arg)
    238 def compute_flux(self, *arg):
    239     """Compute flux"""
--> 240     return self.model.integrate_geom(self.geom, self.gti)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/cube.py:373, in SkyModel.integrate_geom(self, geom, gti, oversampling_factor)
    365 value = self.spectral_model.integral(
    366     energy[:-1],
    367     energy[1:],
    368 ).reshape((-1, 1, 1))
    370 if self.spatial_model:
    371     value = (
    372         value
--> 373         * self.spatial_model.integrate_geom(
    374             geom, oversampling_factor=oversampling_factor
    375         ).quantity
    376     )
    378 if self.temporal_model:
    379     integral = self.temporal_model.integral(gti.time_start, gti.time_stop)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/spatial.py:225, in SpatialModel.integrate_geom(self, geom, oversampling_factor)
    220 if self.evaluation_radius is not None:
    221     # Is it still needed?
    222     width = 2 * np.maximum(
    223         self.evaluation_radius.to_value("deg"), pix_scale
    224     )
--> 225     wcs_geom = wcs_geom.cutout(self.position, width)
    227 upsampled_geom = wcs_geom.upsample(oversampling_factor, axis_name=None)
    229 # assume the upsampled solid angles are approximately factor**2 smaller

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/wcs/geom.py:886, in WcsGeom.cutout(self, position, width, mode, odd_npix)
    883     width = round_up_to_odd(width_npix)
    885 dummy_data = np.empty(self.to_image().data_shape, dtype=bool)
--> 886 c2d = Cutout2D(
    887     data=dummy_data,
    888     wcs=self.wcs,
    889     position=position,
    890     # Cutout2D takes size with order (lat, lon)
    891     size=width[::-1],
    892     mode=mode,
    893 )
    894 return self._init_copy(wcs=c2d.wcs, npix=c2d.shape[::-1])

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/astropy/nddata/utils.py:571, in Cutout2D.__init__(self, data, position, size, wcs, mode, fill_value, copy)
    567 # reverse position because extract_array and overlap_slices
    568 # use (y, x), but keep the input position
    569 pos_yx = position[::-1]
--> 571 cutout_data, input_position_cutout = extract_array(
    572     data, tuple(shape), pos_yx, mode=mode, fill_value=fill_value,
    573     return_position=True)
    574 if copy:
    575     cutout_data = np.copy(cutout_data)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/astropy/nddata/utils.py:216, in extract_array(array_large, shape, position, mode, fill_value, return_position)
    213 if mode not in ['partial', 'trim', 'strict']:
    214     raise ValueError("Valid modes are 'partial', 'trim', and 'strict'.")
--> 216 large_slices, small_slices = overlap_slices(array_large.shape,
    217                                             shape, position, mode=mode)
    218 extracted_array = array_large[large_slices]
    219 if return_position:

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/astropy/nddata/utils.py:109, in overlap_slices(large_array_shape, small_array_shape, position, mode)
    107 for e_max in indices_max:
    108     if e_max < 0:
--> 109         raise NoOverlapError('Arrays do not overlap.')
    110 for e_min, large_shape in zip(indices_min, large_array_shape):
    111     if e_min >= large_shape:

NoOverlapError: Arrays do not overlap.

**Reply by registerrier** [reply]
Hi @astroSyvertHE , this error might be caused by a model with an extended source that lies outside the region covered by the reference `dataset`. This is likely an issue with the `Model` evaluation on a `Dataset`.

To avoid this issue, you can simply select the models that are currently covered by theIf your dataset has a `mask_safe` defined, you can filter the models with e.g. `models.select_mask(dataset.mask_safe)`.

See e.g. https://docs.gammapy.org/1.0/api/gammapy.modeling.models.Models.html#gammapy.modeling.models.Models.select_mask

**Reply by astroSyvertHE** [reply]
Hi,
I tried to solve this issue, but I'm not able to do that. I tried with Models.reassign, but it doesn't work.
I read two yaml models with Models.reads and I want to generate one models with these two + bkg + iem using Models([firtst_model, second_model, bkg, iem]), but it doesn't work because the models read from yaml haven't the dataset name (None). This is the error:

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [128], in <cell line: 9>()
      1 """
      2 gc_src = Models([gc_pwn, gc_snr])
      3 print(gc_src[0])
      4 print(gc_src[-1])
      5 """
      6 #gc_src = gc_pwn.copy()
      7 #gc_src.extend(gc_snr)
      8 #gc_src = gc_pwn + gc_snr
----> 9 gc_src = Models([gc_pwn, gc_snr])
     10 print(gc_src.names)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/core.py:344, in DatasetModels.__init__(self, models)
    342 unique_names = []
    343 for model in models:
--> 344     if model.name in unique_names:
    345         raise (ValueError("Model names must be unique"))
    346     unique_names.append(model.name)

AttributeError: 'Models' object has no attribute 'name'

Could you help me?

Thank you very much

Cheers,

sofiav

**Reply by AtreyeeS** [reply]
Hi @astroSyvertHE ! I think your `gc_src[0]` is not a single `model`, but already a list of `models`.

**Reply by astroSyvertHE** [reply]
Yes, it's a list of models, 70 in the firts and 10 in the second. If I do gc_src = Models([gc_pwn[0], gc_snr[0]), it works, but I want to do that for all the models in the two lists...

**Reply by QRemy** [reply]
Hi,
The error is not related to Datasets names but to Model names, here it seems that `gc_pwn`, `gc_snr`  are `Models`  not `Model`.
 If you want to combine them it should be : 
`gc_src = gc_pwn + gc_snr`

**Reply by AtreyeeS** [reply]
I guess you can also do `Models.extend()` to combine two `Models` objects

**Reply by astroSyvertHE** [reply]
Thank you very much!
I'm following the tutorial of 3D map simulation. Here creates model_simu, but just with one model. Since I have a list of models, can I do the same?

Thank you again

Cheers,

sofiav

**Reply by AtreyeeS** [reply]
Yes indeed, you can set models objects as ``dataset.models`

**Reply by astroSyvertHE** [reply]
I write this:

models = DatasetModels(gc_src_bkg_iem)
type(models)
model_geom = models.wcs_geom
model_spec = SpectralModel.from_parameters(gc_src_bkg_iem)
print(model_geom)
print(model_spec)
model_simu = models.to_template_sky_model(model_geom, spectral_model = model_spec, name = "model-simu")

and this is the output:

WcsGeom

	axes       : ['lon', 'lat']
	shape      : (2082, 1800)
	ndim       : 2
	frame      : icrs
	projection : TAN
	center     : 266.4 deg, -28.9 deg
	width      : 208.2 deg x 180.0 deg
	wcs ref    : 266.4 deg, -28.9 deg

SpectralModel

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Input In [214], in <cell line: 7>()
      5 print(model_geom)
      6 print(model_spec)
----> 7 model_simu = models.to_template_sky_model(model_geom, spectral_model = model_spec, name = "model-simu")

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/core.py:929, in DatasetModels.to_template_sky_model(self, geom, spectral_model, name)
    926 map_ = Map.from_geom(geom, unit=unit)
    928 for m in self:
--> 929     map_ += m.evaluate_geom(geom).to(unit)
    931 spatial_model = TemplateSpatialModel(map_, normalize=False)
    933 if spectral_model is None:

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/cube.py:332, in SkyModel.evaluate_geom(self, geom, gti)
    330 """Evaluate model on `~gammapy.maps.Geom`."""
    331 coords = geom.get_coord(sparse=True)
--> 332 value = self.spectral_model(coords["energy_true"])
    334 if self.spatial_model:
    335     value = value * self.spatial_model.evaluate_geom(geom)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/coord.py:55, in MapCoord.__getitem__(self, key)
     53 def __getitem__(self, key):
     54     if isinstance(key, str):
---> 55         return self._data[key]
     56     else:
     57         return list(self._data.values())[key]

KeyError: 'energy_true'

Seems that is not the correct way to select the spectral models in the model list too....

You are angels if you help me!

Thank you again!

sofiav

**Reply by AtreyeeS** [reply]
The issue here is that your `geom` does not have an `energy_true` axis. Why do you want to use `to_template_sky_model`? You can do 
```
models1 = Models.read(...)
model12 = Models.read(...)
models1.extend(models2)
dataset.models = models1
dataset.fake()
```

**Reply by astroSyvertHE** [reply]
How can I compute the flux from a rectangular region?

Is it possible to use a rectangular region in the spatial model definition?

Thank you

sofiav

**Reply by QRemy** [reply]
No, there is no rectangular spatial model but you can create a mask from a rectangular region (as shown [here](https://docs.gammapy.org/1.0/tutorials/api/mask_maps.html#create-the-mask-from-a-list-of-regions)) and evaluate the flux only where the mask is true.
Why do you need a rectangle region, what's the use case ?

**Reply by astroSyvertHE** [reply]
I have to check if a population of synthetic sources are compatible with observed flux in a rectangular region. So at this preliminary point I want to estimate the flux from this region with different exposures. I need ccube and expcube to do that by myself, but I don't understand how to have these two files with gammapy. Or eventually if it is possible to do that with gammapy classes....

**Reply by astroSyvertHE** [reply]
Please, what is the solution of the exercise in the Data Structure turorial?
Make a counts energy spectrum for the galactic center region, within a radius of 10 deg.

Thanks

sofiav

**Reply by AtreyeeS** [reply]
- For your use case of computing the flux in a rectangular region, you can do exactly what @QRemy suggested.
- If you are doing a Fermi analysis, the data reduction has to be done outside gammapy. You can use Fermi Science tools or fermipy to compute the cube and expcube and then do the modelling and fitting inside gammapy

**Reply by astroSyvertHE** [reply]
I just performed a simulation of these sources with CTA south, and I have the EventList.

Please, what is the solution of the exercise in the Data Structure turorial?
Make a counts energy spectrum for the galactic center region, within a radius of 10 deg.

Thanks

sofiav

**Reply by astroSyvertHE** [reply]
Apparently I'm using the same code of the tutorials... but there's this error, why?

AttributeError                            Traceback (most recent call last)
Input In [517], in <cell line: 3>()
      1 datasets = Datasets()
----> 3 dataset_on = dataset_maker.run(
      4         dataset_empty.copy(name="20230313"), observation
      5 )
      6 dataset_on_off = bkg_maker.run(dataset_on, observation)
      7 dataset_on_off = safe_mask_masker.run(dataset_on_off, observation)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/makers/map.py:389, in MapDatasetMaker.run(self, dataset, observation)
    386 kwargs["mask_safe"] = mask_safe
    388 if "counts" in self.selection:
--> 389     counts = self.make_counts(dataset.counts.geom, observation)
    390 else:
    391     counts = Map.from_geom(dataset.counts.geom, data=0)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/makers/map.py:151, in MapDatasetMaker.make_counts(geom, observation)
    149 else:
    150     counts = Map.from_geom(geom)
--> 151     counts.fill_events(observation.events)
    152 return counts

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/core.py:1091, in Map.fill_events(self, events)
   1089 def fill_events(self, events):
   1090     """Fill event coordinates (`~gammapy.data.EventList`)."""
-> 1091     self.fill_by_coord(events.map_coord(self.geom))

AttributeError: 'NoneType' object has no attribute 'map_coord'


Thanks

sofiav

**Reply by AtreyeeS** [reply]
It seems you do not have `events` set on your observation. `observation.events` is None

```
# copy the observation in memory, setting the events
o2 = obs.copy(events=events_new, in_memory=True)

# The new observation and the new events table can be serialised independently
o2.write("new_obs.fits.gz")
events_new.write("events.fits.gz", gti=obs.gti, overwrite=True)
```

**Reply by astroSyvertHE** [reply]
I simulate the dataset, I'm not using fits file. In the tutorials dataStore are created with observations. But in my case, in particular what I have to use? I have the EventList make with Maker, and dataset that's a Map....

Thanks,

sofiav

**Reply by QRemy** [reply]
> Please, what is the solution of the exercise in the Data Structure turorial? Make a counts energy spectrum for the galactic center region, within a radius of 10 deg.
> 

```
from regions import Regions
from gammapy.maps import RegionGeom
regions_ds9 = "galactic;circle(0,0,10)"
regions = Regions.parse(regions_ds9, format="ds9")
region_geom = RegionGeom(regions[0])
dataset.counts.to_region_nd_map(region_geom).plot()
```

**Reply by astroSyvertHE** [reply]
There's this issue with ON-OFF analysis. The on_region is defined as a circular region like in tutorials.

ValueError                                Traceback (most recent call last)
Input In [942], in <cell line: 12>()
      8 #stacked_dataset.models = model
      9 #on_region = [on_region1, on_region2, on_region3, on_region4, on_region5, on_region6]
     11 fpe = FluxPointsEstimator(energy_edges=energy_edges, source="on_region")
---> 12 flux_points = fpe.run(datasets=[stacked_dataset])
     13 flux_points.to_table(sed_type="dnde", formatted=True)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/estimators/points/sed.py:113, in FluxPointsEstimator.run(self, datasets)
    109 rows = []
    110 for energy_min, energy_max in progress_bar(
    111     zip(self.energy_edges[:-1], self.energy_edges[1:]), desc="Energy bins"
    112 ):
--> 113     row = self.estimate_flux_point(
    114         datasets,
    115         energy_min=energy_min,
    116         energy_max=energy_max,
    117     )
    118     rows.append(row)
    120 meta = {
    121     "n_sigma": self.n_sigma,
    122     "n_sigma_ul": self.n_sigma_ul,
    123     "sed_type_init": "likelihood",
    124 }

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/estimators/points/sed.py:160, in FluxPointsEstimator.estimate_flux_point(self, datasets, energy_min, energy_max)
    158 if len(datasets_sliced) > 0:
    159     datasets_sliced.models = datasets.models.copy()
--> 160     return super().run(datasets=datasets_sliced)
    161 else:
    162     log.warning(f"No dataset contribute in range {energy_min}-{energy_max}")

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/estimators/flux.py:177, in FluxEstimator.run(self, datasets)
    174 datasets = Datasets(datasets)
    175 models = datasets.models.copy()
--> 177 model = self.get_scale_model(models)
    179 energy_min, energy_max = datasets.energy_ranges
    180 energy_axis = MapAxis.from_energy_edges([energy_min.min(), energy_max.max()])

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/estimators/flux.py:115, in FluxEstimator.get_scale_model(self, models)
    102 def get_scale_model(self, models):
    103     """Set scale model
    104 
    105     Parameters
   (...)
    113         Scale spectral model
    114     """
--> 115     ref_model = models[self.source].spectral_model
    117     if ref_model.is_norm_spectral_model:
    118         raise ValueError(
    119             "Instances of `NormSpectralModel` are not supported for flux point estimation."
    120         )

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/core.py:617, in DatasetModels.__getitem__(self, key)
    615     return self.__class__(list(np.array(self._models)[key]))
    616 else:
--> 617     return self._models[self.index(key)]

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/core.py:623, in DatasetModels.index(self, key)
    621     return key
    622 elif isinstance(key, str):
--> 623     return self.names.index(key)
    624 elif isinstance(key, ModelBase):
    625     return self._models.index(key)

ValueError: 'on_region' is not in list

**Reply by astroSyvertHE** [reply]
Hi,
I create a mask as here:

estimator = ExcessMapEstimator(
    correlation_radius="0.1 deg",
    selection_optional=[],
    energy_edges=[0.1, 1, 10] * u.TeV,
)

result = estimator.run(dataset_on)
result["sqrt_ts"].plot_grid(
    figsize=(12, 4), cmap="coolwarm", add_cbar=True, vmin=-5, vmax=5, ncols=2
)
mask_on = result["sqrt_ts"] < 5.0
invalid_pixels = np.isnan(result["sqrt_ts"].data)
mask_on.data[invalid_pixels] = True

mask_on.plot_interactive()


Now I want to use this mask as spatial model for fitting dataset. I'm using TemplateSpatialModel(mask_on), but it doesn't work...
Could you help me?
Thank you!

sofiav

**Reply by AtreyeeS** [reply]
Hi Sofia,

I dont understand your use case. Why do you want to create a model from a mask? What error are you getting when you say it does not work?

**Reply by astroSyvertHE** [reply]
with a CircleSkyRegion I used TemplateSpatialModel to define the spatial model to use in model=SkyModel(spectral_model, spatial_model)

Now I want to use the musk estimated from excesses to generate the spatial model. I did that:

template_on2 = TemplateSpatialModel(mask_on)#.to_region_nd_map())#.sum_over_axes())
#print(type(template_on2))
on_model2 = SkyModel(
   spectral_model= PowerLawSpectralModel(index=2, amplitude="1e-11 cm-2 s-1 TeV-1", reference="1 TeV"),
   spatial_model=template_on2,
   name="on_region",
)

and the error is:
KeyError                                  Traceback (most recent call last)
Input In [587], in <cell line: 4>()
      2 template_on2 = TemplateSpatialModel(mask_on)#.to_region_nd_map())#.sum_over_axes())
      3 #print(type(template_on2))
----> 4 on_model2 = SkyModel(
      5    spectral_model= PowerLawSpectralModel(index=2, amplitude="1e-11 cm-2 s-1 TeV-1", reference="1 TeV"),
      6    spatial_model=template_on2,
      7    name="on_region",
      8 )

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/cube.py:72, in SkyModel.__init__(self, spectral_model, spatial_model, temporal_model, name, apply_irf, datasets_names)
     70 self.apply_irf = apply_irf
     71 self.datasets_names = datasets_names
---> 72 self._check_unit()
     74 is_norm = np.array([par.is_norm for par in spectral_model.parameters])
     76 if not np.any(is_norm):

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/cube.py:106, in SkyModel._check_unit(self)
    103 geom = WcsGeom.create(skydir=self.position, npix=(2, 2), axes=[axis])
    105 gti = GTI.create(1 * u.day, 2 * u.day)
--> 106 value = self.evaluate_geom(geom, gti)
    108 if self.apply_irf["exposure"]:
    109     ref_unit = u.Unit("cm-2 s-1 MeV-1 sr-1")

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/cube.py:335, in SkyModel.evaluate_geom(self, geom, gti)
    332 value = self.spectral_model(coords["energy_true"])
    334 if self.spatial_model:
--> 335     value = value * self.spatial_model.evaluate_geom(geom)
    337 if self.temporal_model:
    338     integral = self.temporal_model.integral(gti.time_start, gti.time_stop)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/spatial.py:172, in SpatialModel.evaluate_geom(self, geom)
    170     return self(coords.lon, coords.lat, energy=coords["energy_true"])
    171 else:
--> 172     return self(coords.lon, coords.lat)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/spatial.py:80, in SpatialModel.__call__(self, lon, lat, energy)
     77 if energy is not None:
     78     kwargs["energy"] = energy
---> 80 return self.evaluate(lon, lat, **kwargs)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/modeling/models/spatial.py:1153, in TemplateSpatialModel.evaluate(self, lon, lat, energy)
   1150 if energy is not None:
   1151     coord["energy_true"] = energy
-> 1153 val = self.map.interp_by_coord(coord, **self._interp_kwargs)
   1154 val = np.clip(val, 0, a_max=None)
   1155 return u.Quantity(val, self.map.unit, copy=False)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/wcs/ndmap.py:132, in WcsNDMap.interp_by_coord(self, coords, method, fill_value)
    129 def interp_by_coord(self, coords, method="linear", fill_value=None):
    131     if self.geom.is_regular:
--> 132         pix = self.geom.coord_to_pix(coords)
    133         return self.interp_by_pix(pix, method=method, fill_value=fill_value)
    134     else:

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/wcs/geom.py:638, in WcsGeom.coord_to_pix(self, coords)
    635 else:
    636     pix = self._wcs.wcs_world2pix(coords.lon, coords.lat, 0)
--> 638 pix += self.axes.coord_to_pix(coords)
    639 return tuple(pix)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/axes.py:1753, in MapAxes.coord_to_pix(self, coord)
   1740 def coord_to_pix(self, coord):
   1741     """Transform from axis to pixel coordinates.
   1742 
   1743     Parameters
   (...)
   1751         Array of pixel coordinate values.
   1752     """
-> 1753     return tuple([ax.coord_to_pix(coord[ax.name]) for ax in self])

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/axes.py:1753, in <listcomp>(.0)
   1740 def coord_to_pix(self, coord):
   1741     """Transform from axis to pixel coordinates.
   1742 
   1743     Parameters
   (...)
   1751         Array of pixel coordinate values.
   1752     """
-> 1753     return tuple([ax.coord_to_pix(coord[ax.name]) for ax in self])

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/maps/coord.py:55, in MapCoord.__getitem__(self, key)
     53 def __getitem__(self, key):
     54     if isinstance(key, str):
---> 55         return self._data[key]
     56     else:
     57         return list(self._data.values())[key]

KeyError: 'energy'

**Reply by registerrier** [reply]
Hi @astroSyvertHE , the `TemplateSpatialModel` does not support a `Map` with `energy` axis. It can support an `energy_true` axis or no non-spatial axis at all. For this to work you have, for instance, to remove the energy axis.

Yet, as @AtreyeeS asked, it seems a bit unclear what fitting a model built from the map of non-significant features can yield.

**Reply by astroSyvertHE** [reply]
Hi,
I want to analize extended source with RectangleSkyRegion using reflected backround, but there's the error I'm not able to figure out:
WARNING:gammapy.makers.background.reflected:ReflectedRegionsBackgroundMaker failed. No OFF region found outside exclusion mask for dataset 'obs-20230324'.
WARNING:gammapy.makers.background.reflected:ReflectedRegionsBackgroundMaker failed. Setting obs-20230324 mask to False.

the code is:
# The binning of the final spectrum is defined here.
energy_axis = MapAxis.from_energy_bounds(0.1, 40.0, 10, unit="TeV")

# Reduced IRFs are defined in true energy (i.e. not measured energy).
energy_axis_true = MapAxis.from_energy_bounds(
    0.05, 100, 30, unit="TeV", name="energy_true"
)

geom_on = RegionGeom.create(ridge_hess, axes=[energy_axis])

dataset_empty = SpectrumDataset.create(geom=geom_on, energy_axis_true=energy_axis_true)

dataset_maker = SpectrumDatasetMaker(
    selection=["counts", "exposure", "edisp"],
    use_region_center=False
)


finder = ReflectedRegionsFinder(min_distance_input='0.05 deg', max_region_number = 3)
center_pointing = SkyCoord(0,1, 0.02, unit='deg', frame='galactic')
regions, wcs = finder.run(region=ridge_hess, center=center_pointing, exclusion_mask=exclusion_ridge)
print(regions[0])
#print(wcs)

bkg_maker = ReflectedRegionsBackgroundMaker()#region_finder=regions[0], exclusion_mask=exclusion_ridge)
#safe_mask_masker = SafeMaskMaker(methods=["aeff-max"], aeff_percent=10)#10
safe_mask_maker = SafeMaskMaker(methods=["aeff-max"], aeff_percent=10)

dataset_on = dataset_maker.run(
        dataset_empty.copy(name="obs-20230324"), o2, 
)
dataset_on = safe_mask_maker.run(dataset_on, o2)

dataset_on = bkg_maker.run(dataset_on, o2)
#dataset_on_off = safe_mask_masker.run(dataset_on_off, o2)
#datasets.append(dataset_on)
display(dataset_on.meta_table)


Could you help me? thank you!

Cheers,

sofiav

**Reply by registerrier** [reply]
Here, what is happening most likely is that no reflected regions have been found because the size of your on region is too wide. If you are using a box of more than a degree across (as needed to cover the HESS GC ridge emission), it is likely that no reflected regions can fit without a very large pointing offset from the GC.
I think in this case, the reflected region is probably not the adapted method for such an analysis.

**Reply by astroSyvertHE** [reply]
Thank you.
I just performed the 3D analysis, but I'd like to use another method to see if the result I found is consistenti... So, the reflected background is not usable because the region size is to big. I suppose that also the ring wooble method is not useful, is it?

Thanks,

sofiav

**Reply by registerrier** [reply]
The ring background is usually used to correct the background of an image (not a cube). The wobble OFF region is similar to the reflected region but simply select regions at specific locations (namely symmetric positions w.r.t. FoV center). It will suffer from the same difficulties as the reflected regions.
For now, there is no background Maker except for the FoV background normalization for a case like this. But it does not allow extracting OFF counts to constrain the background.

**Reply by astroSyvertHE** [reply]
Hi,
I create an events sampling using the tutorial /analysis_3d/events_sampling, and in particular mutiple events sampling. Next, I used /starting/analysis_2 tutorial.

The error is:
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Input In [36], in <cell line: 1>()
      3 cutout = dataset_stacked.cutout(
      4     obs.pointing_radec, width=2 * offset_max, name=f"obs-{obs.obs_id}"
      5 )
      6 # A MapDataset is filled in this cutout geometry
----> 7 dataset = maker.run(cutout, obs)
      8 # The data quality cut is applied
      9 dataset = maker_safe_mask.run(dataset, obs)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/makers/map.py:395, in MapDatasetMaker.run(self, dataset, observation)
    392 kwargs["counts"] = counts
    394 if "exposure" in self.selection:
--> 395     exposure = self.make_exposure(dataset.exposure.geom, observation)
    396     kwargs["exposure"] = exposure
    398 if "background" in self.selection:

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/makers/map.py:176, in MapDatasetMaker.make_exposure(geom, observation, use_region_center)
    170 if isinstance(observation.aeff, Map):
    171     return observation.aeff.interp_to_geom(
    172         geom=geom,
    173     )
    174 return make_map_exposure_true_energy(
    175     pointing=observation.pointing_radec,
--> 176     livetime=observation.observation_live_time_duration,
    177     aeff=observation.aeff,
    178     geom=geom,
    179     use_region_center=use_region_center,
    180 )

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/data/observations.py:255, in Observation.observation_live_time_duration(self)
    246 @property
    247 def observation_live_time_duration(self):
    248     """Live-time duration in seconds (`~astropy.units.Quantity`).
    249 
    250     The dead-time-corrected observation time.
   (...)
    253     where ``f_dead`` is the dead-time fraction.
    254     """
--> 255     return self.observation_time_duration * (
    256         1 - self.observation_dead_time_fraction
    257     )

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/data/observations.py:244, in Observation.observation_time_duration(self)
    238 @property
    239 def observation_time_duration(self):
    240     """Observation time duration in seconds (`~astropy.units.Quantity`).
    241 
    242     The wall time, including dead-time.
    243     """
--> 244     return self.gti.time_sum

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/data/gti.py:170, in GTI.time_sum(self)
    167 @property
    168 def time_sum(self):
    169     """Sum of GTIs in seconds (`~astropy.units.Quantity`)."""
--> 170     return self.time_delta.sum()

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/gammapy/data/gti.py:158, in GTI.time_delta(self)
    155 @property
    156 def time_delta(self):
    157     """GTI durations in seconds (`~astropy.units.Quantity`)."""
--> 158     start = self.table["START"].astype("float64")
    159     stop = self.table["STOP"].astype("float64")
    160     return u.Quantity(stop - start, "second")

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/astropy/table/table.py:1874, in Table.__getitem__(self, item)
   1872 def __getitem__(self, item):
   1873     if isinstance(item, str):
-> 1874         return self.columns[item]
   1875     elif isinstance(item, (int, np.integer)):
   1876         return self.Row(self, item)

File ~/miniconda3/envs/gammapy-1.0/lib/python3.9/site-packages/astropy/table/table.py:246, in TableColumns.__getitem__(self, item)
    236 """Get items from a TableColumns object.
    237 ::
    238 
   (...)
    243   tc[1:3] # <TableColumns names=('b', 'c')>
    244 """
    245 if isinstance(item, str):
--> 246     return OrderedDict.__getitem__(self, item)
    247 elif isinstance(item, (int, np.integer)):
    248     return list(self.values())[item]

KeyError: 'START'


I used the same code lines of the two tutorials....

Thanks,

Cheers,

sofiav
